cfn-resources/project/cmd/resource/resource.go

```diff
@@ -467,75 +485,75 @@ func readProjectSettings(client *mongodbatlas.Client, id string, currentModel *M
+	changedTeams []admin.TeamRole, removeTeams []admin.TeamRole) {
 	for _, nTeam := range currentTeams {
 		if nTeam.TeamId != nil && len(*nTeam.TeamId) > 0 {
 			matched := false
 			for _, oTeam := range oTeams {
-				if nTeam.TeamId != nil && *nTeam.TeamId == oTeam.TeamID {
-					changedTeams = append(changedTeams, &mongodbatlas.ProjectTeam{TeamID: *nTeam.TeamId, RoleNames: nTeam.RoleNames})
+				if nTeam.TeamId != nil && oTeam.TeamId != nil && *nTeam.TeamId == *oTeam.TeamId {
+					changedTeams = append(changedTeams, admin.TeamRole{TeamId: nTeam.TeamId, RoleNames: nTeam.RoleNames})
 					matched = true
 					break
 				}
 			}
 			// Add to newTeams
 			if !matched {
-				newTeams = append(newTeams, &mongodbatlas.ProjectTeam{TeamID: *nTeam.TeamId, RoleNames: nTeam.RoleNames})
+				newTeams = append(newTeams, admin.TeamRole{TeamId: nTeam.TeamId, RoleNames: nTeam.RoleNames})
 			}
 		}
 	}
 
 	for _, oTeam := range oTeams {
-		if len(oTeam.TeamID) > 0 {
+		if oTeam.TeamId != nil && len(*oTeam.TeamId) > 0 {
 			matched := false
 			for _, nTeam := range currentTeams {
-				if nTeam.TeamId != nil && *nTeam.TeamId == oTeam.TeamID {
+				if nTeam.TeamId != nil && oTeam.TeamId != nil && *nTeam.TeamId == *oTeam.TeamId {
 					matched = true
 					break
 				}
 			}
 			if !matched {
-				removeTeams = append(removeTeams, &mongodbatlas.ProjectTeam{TeamID: oTeam.TeamID, RoleNames: oTeam.RoleNames})
+				removeTeams = append(removeTeams, admin.TeamRole{TeamId: oTeam.TeamId, RoleNames: oTeam.RoleNames})
 			}
 		}
 	}
 	return newTeams, changedTeams, removeTeams
 }
 
 // Get difference in ApiKeys
-func getChangeInAPIKeys(groupID string, currentKeys []ProjectApiKey, oKeys []mongodbatlas.APIKey) (newKeys, changedKeys, removeKeys []UpdateAPIKey) {
+func getChangeInAPIKeys(groupID string, currentKeys []ProjectApiKey, oKeys []admin.ApiKeyUserDetails) (newKeys, changedKeys, removeKeys []UpdateAPIKey) {
 	for _, nKey := range currentKeys {
 		if nKey.Key != nil && len(*nKey.Key) > 0 {
 			matched := false
 			for _, oKey := range oKeys {
-				if nKey.Key != nil && *nKey.Key == oKey.ID {
-					changedKeys = append(changedKeys, UpdateAPIKey{Key: *nKey.Key, APIKeys: &mongodbatlas.AssignAPIKey{Roles: nKey.RoleNames}})
+				if nKey.Key != nil && oKey.Id != nil && *nKey.Key == *oKey.Id {
+					changedKeys = append(changedKeys, UpdateAPIKey{Key: *nKey.Key, UpdatePayload: &admin.UpdateAtlasProjectApiKey{Roles: nKey.RoleNames}})
 					matched = true
 					break
 				}
 			}
 			// Add to newKeys
 			if !matched {
-				newKeys = append(newKeys, UpdateAPIKey{Key: *nKey.Key, APIKeys: &mongodbatlas.AssignAPIKey{Roles: nKey.RoleNames}})
+				newKeys = append(newKeys, UpdateAPIKey{Key: *nKey.Key, UpdatePayload: &admin.UpdateAtlasProjectApiKey{Roles: nKey.RoleNames}})
 			}
 		}
 	}
 
 	for _, oKey := range oKeys {
-		if len(oKey.ID) > 0 {
+		if oKey.Id != nil && len(*oKey.Id) > 0 {
```

Comment:
i see this pattern a lot, not sure if we could have a helper method, e.g. like in TF:

func IsStringPresent(strPtr *string) bool {
	return strPtr != nil && len(*strPtr) > 0
}

and also for comparison like 

func AreStringPtrEqual(p1, p2 *string) bool {
    if  p1 == nil  {
       return p2 == nil;
   }
   if (p2 == nil) {
      return false;
  }
 return *p1 == *p2
}


maybe this could be a good "excuse" / opportunity to have a common repo for TF / CFN / CDK / etc. if we envision to have some common functions.

Reply:
As a starting point extracted these common functions. Like the idea of the common repo, but we will have to see if it justifies for the amount of common code (taking into account that some helper functions are already being placed in sdk).